aten/src/ATen/native/cuda/layer_norm_kernel.cu

```diff
@@ -27,23 +27,37 @@ __global__ void RowwiseMomentsCUDAKernel(
     T* mean,
     T* rstd) {
   using T_ACC = acc_type<T, true>;
+
   __shared__ T_ACC m_shared[C10_WARP_SIZE];
   __shared__ T_ACC v_shared[C10_WARP_SIZE];
+  __shared__ T_ACC mean_;
+
   const int64_t i = blockIdx.x;
+  const T_ACC scale = T_ACC(1) / static_cast<T_ACC>(N);
+
   T_ACC sum1 = 0;
-  T_ACC sum2 = 0;
   for (int64_t j = threadIdx.x; j < N; j += blockDim.x) {
     const int64_t index = i * N + j;
     sum1 += static_cast<T_ACC>(X[index]);
-    sum2 += static_cast<T_ACC>(X[index]) * static_cast<T_ACC>(X[index]);
   }
   sum1 = cuda_utils::BlockReduceSum<T_ACC>(sum1, m_shared);
+
+  if (threadIdx.x == 0) {
+      mean_ = sum1 * scale;
+  }
+  __syncthreads();
```

Comment:
Is syncing actually faster than multiplying on each thread separately?

Reply:
The problem here is that after `BlockReduceSum`, only the first thread in the block has the fully reduced value of `sum1`. So, in my understanding, we would need to sync anyway in order to broadcast `sum1` to all threads in the block.